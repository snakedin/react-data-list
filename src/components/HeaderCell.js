import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { withOptions } from "../options/withOptions";

class HeaderCell extends Component {

    static propTypes = {
        id: PropTypes.string,
        label: PropTypes.string,
        attributes: PropTypes.object,
        onChangeSort: PropTypes.func,
        currentSort: PropTypes.string
    };

    /**
     *  Parse column header label
     */
    parseLabel() {
        const { id, label } = this.props;

        if (label !== undefined && typeof label === "string") {
            return label;
        } else {
            return id.charAt(0).toUpperCase() + id.slice(1);
        }
    }

    /**
     * Render cell content
     */
    renderCell() {
        const { id, sort, onChangeSort, currentSort, options: { enableSorting } } = this.props;

        let content = this.parseLabel();

        if (sort !== false && enableSorting) {

            let sortBy = sort;
            let className = `btn btn-sm btn-link ${currentSort.charAt(0) === '-' ? 'desc' : 'asc'}`;

            if (sortBy === true || sortBy === undefined) {
                sortBy = id;
            }

            if ((currentSort.charAt(0) === '-' && currentSort.slice(1) === sortBy) || currentSort === sortBy) {
                className += ' active';
            }

            content = (
                <span onClick={ () => { onChangeSort(sortBy) } } className={className}>{ content }</span>
            )
        } else {
            content = <span className="btn btn-sm">{content}</span>;
        }

        return content;
    }

    render() {
        const { attributes } = this.props;

        return (
            <th scope="col" {...attributes}>
                { this.renderCell() }
            </th>
        );
    }
}

export default withOptions(HeaderCell);