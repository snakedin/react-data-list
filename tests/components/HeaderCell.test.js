import React from "react";
import { render } from "@testing-library/react";
import "../mocks/withOptionsMock";
import HeaderCell from "../../src/components/HeaderCell";

it("HeaderCell render", () => {

  function thRender(component) {
    const tableRow = document.createElement('tr');

    const { container } = render(component, {
      container: document.body.appendChild(tableRow)
    });

    return container;
  }

  const baseRender = thRender(<HeaderCell id="name" currentSort={''}/>);
  expect(baseRender).toMatchSnapshot();

  expect(baseRender).toMatchDiffSnapshot(
      thRender(<HeaderCell id="name" attributes={{'className': 'header-cell'}} currentSort={''}/>)
  );

  expect(baseRender).toMatchDiffSnapshot(
      thRender(<HeaderCell id="name" label="User Name" currentSort={''}/>)
  );

  expect(baseRender).toMatchDiffSnapshot(
      thRender(<HeaderCell id="name" sort={false} currentSort={''}/>)
  );

  expect(baseRender).toMatchDiffSnapshot(
      thRender(<HeaderCell id="name" sort={true} currentSort={"name"}/>)
  );

  expect(baseRender).toMatchDiffSnapshot(
      thRender(<HeaderCell id="name" sort={true} currentSort={"-name"}/>)
  );

  expect(baseRender).toMatchDiffSnapshot(
      thRender(<HeaderCell id="name" sort={"userName"} currentSort={"userName"}/>)
  );
});
