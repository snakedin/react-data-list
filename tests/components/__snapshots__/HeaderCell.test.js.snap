// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeaderCell render 1`] = `
<tr>
  <th
    scope="col"
  >
    <span
      class="btn btn-sm btn-link asc"
    >
      Name
    </span>
  </th>
</tr>
`;

exports[`HeaderCell render 2`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -1,7 +1,8 @@
  <tr>
    <th
+     class=\\"header-cell\\"
      scope=\\"col\\"
    >
      <span
        class=\\"btn btn-sm btn-link asc\\"
      >"
`;

exports[`HeaderCell render 3`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -3,9 +3,9 @@
      scope=\\"col\\"
    >
      <span
        class=\\"btn btn-sm btn-link asc\\"
      >
-       Name
+       User Name
      </span>
    </th>
  </tr>"
`;

exports[`HeaderCell render 4`] = `
"Snapshot Diff:
- First value
+ Second value

  <tr>
    <th
      scope=\\"col\\"
    >
      <span
-       class=\\"btn btn-sm btn-link asc\\"
+       class=\\"btn btn-sm\\"
      >
        Name
      </span>
    </th>
  </tr>"
`;

exports[`HeaderCell render 5`] = `
"Snapshot Diff:
- First value
+ Second value

  <tr>
    <th
      scope=\\"col\\"
    >
      <span
-       class=\\"btn btn-sm btn-link asc\\"
+       class=\\"btn btn-sm btn-link asc active\\"
      >
        Name
      </span>
    </th>
  </tr>"
`;

exports[`HeaderCell render 6`] = `
"Snapshot Diff:
- First value
+ Second value

  <tr>
    <th
      scope=\\"col\\"
    >
      <span
-       class=\\"btn btn-sm btn-link asc\\"
+       class=\\"btn btn-sm btn-link desc active\\"
      >
        Name
      </span>
    </th>
  </tr>"
`;

exports[`HeaderCell render 7`] = `
"Snapshot Diff:
- First value
+ Second value

  <tr>
    <th
      scope=\\"col\\"
    >
      <span
-       class=\\"btn btn-sm btn-link asc\\"
+       class=\\"btn btn-sm btn-link asc active\\"
      >
        Name
      </span>
    </th>
  </tr>"
`;
